diff --git a/build/CMakeLists.txt b/build/CMakeLists.txt
index 0f561ab..d85b220 100644
--- a/build/CMakeLists.txt
+++ b/build/CMakeLists.txt
@@ -1,4 +1,4 @@
-add_library(librpmbuild SHARED)
+add_library(librpmbuild STATIC)
 set_target_properties(librpmbuild PROPERTIES
 	VERSION ${RPM_LIBVERSION}
 	SOVERSION ${RPM_SOVERSION}
@@ -57,4 +57,4 @@ if (OpenMP_C_FOUND)
 	target_link_libraries(librpmbuild PRIVATE OpenMP::OpenMP_CXX)
 endif()
 
-install(TARGETS librpmbuild EXPORT rpm-targets)
+install(TARGETS librpmbuild libmisc EXPORT rpm-targets)
