diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7c1a9ff..ab61008 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -78,7 +78,6 @@ function(findutil UTIL TRY)
 		PATHS ENV MYPATH
 		PATHS /usr/local/bin /usr/bin /bin
 		PATHS /usr/local/sbin /usr/sbin /sbin
-		NO_DEFAULT_PATH
 	)
 	if (NOT ${UTIL})
 		list(GET TRY 0 util)
@@ -157,8 +156,8 @@ function(makemacros)
 
 	list(GET db_backends 0 DB_BACKEND)
 
-	set(host_cpu ${CMAKE_HOST_SYSTEM_PROCESSOR})
-	string(TOLOWER ${CMAKE_HOST_SYSTEM_NAME} host_os)
+	set(host_cpu "s390x")
+	string(TOLOWER "zos" host_os)
 	set(host_vendor ${RPM_VENDOR})
 	set(host ${host_cpu}-${host_vendor}-${host_os})
 
@@ -196,7 +195,7 @@ set(REQFUNCS
 	mkstemp getcwd basename dirname realpath setenv unsetenv regcomp
 	utimes getline localtime_r statvfs getaddrinfo
 	openat mkdirat fstatat linkat symlinkat mkfifoat mknodat unlinkat
-	renameat utimensat fchmodat fchownat stpcpy stpncpy
+	renameat utimensat fchmodat fchownat
 )
 
 find_package(PkgConfig REQUIRED)
@@ -391,7 +390,7 @@ if (LIBDW_FOUND)
 	set(HAVE_LIBDW 1)
 	check_library_exists(dw dwelf_elf_begin "" HAVE_DWELF_ELF_BEGIN)
 endif()
-
+set(HAVE_SETPROGNAME 1)
 check_symbol_exists(GLOB_ONLYDIR "glob.h" HAVE_GLOB_ONLYDIR)
 check_symbol_exists(major "sys/sysmacros.h" MAJOR_IN_SYSMACROS)
 if (NOT MAJOR_IN_SYSMACROS)
@@ -553,7 +552,7 @@ install(FILES
 	DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/rpm
 )
 
-export(TARGETS librpm librpmio librpmbuild librpmsign
+export(TARGETS librpm librpmio librpmbuild librpmsign libmisc
 	FILE rpm-targets.cmake
 	NAMESPACE rpm::
 )
