diff --git a/build/files.cc b/build/files.cc
index ab14c77..3e4283a 100644
--- a/build/files.cc
+++ b/build/files.cc
@@ -24,7 +24,7 @@
 #include <libelf.h>
 #include <elfutils/libdwelf.h>
 #endif
-
+#include <inttypes.h>
 #include <rpm/argv.h>
 #include <rpm/rpmfc.h>
 #include <rpm/rpmfileutil.h>	/* rpmDoDigest() */
@@ -95,6 +95,7 @@ struct FileListRec_s {
 #define	fl_rdev	fl_st.st_rdev
 #define	fl_size	fl_st.st_size
 #define	fl_mtime fl_st.st_mtime
+#define	fl_genvalue fl_st.st_genvalue
 
     char *diskPath;		/* get file from here       */
     char *cpioPath;		/* filename in cpio archive */
@@ -1101,6 +1102,10 @@ static void genCpioListAndHeader(FileList fl, rpmSpec spec, Package pkg, int isS
     std::sort(fl->files.begin(), fl->files.end());
     
     pkg->dpaths = (char **)xmalloc((fl->files.size() + 1) * sizeof(*pkg->dpaths));
+    std::vector<uint16_t> allModes;
+    std::vector<uint32_t> allGenValues;
+    std::vector<uint16_t> allCCSIDs;
+    std::vector<uint16_t>  allTextFlags;
 
     /* Generate the header. */
     for (i = 0, flp = fl->files.data(); i < fl->files.size(); i++, flp++) {
@@ -1112,7 +1117,6 @@ static void genCpioListAndHeader(FileList fl, rpmSpec spec, Package pkg, int isS
 
 	    /* Two entries for the same file found, merge the entries. */
 	    /* Note that an %exclude is a duplication of a file reference */
-
 	    /* file flags */
 	    flp[1].flags |= flp->flags;	
 
@@ -1185,6 +1189,14 @@ static void genCpioListAndHeader(FileList fl, rpmSpec spec, Package pkg, int isS
 	    rpm_off_t rsize32 = (rpm_off_t)flp->fl_size;
 	    headerPutUint32(h, RPMTAG_FILESIZES, &rsize32, 1);
 	}
+
+	rpm_mode_t linux_mode = (rpm_mode_t) to_linux_mode(flp->fl_mode);
+
+	allModes.push_back(linux_mode);
+	allGenValues.push_back((uint32_t)flp->fl_genvalue);
+	allCCSIDs.push_back(flp->fl_st.st_tag.ft_ccsid);
+	allTextFlags.push_back((uint16_t)(flp->fl_st.st_tag.ft_txtflag ? 1 : 0));
+
 	/* Excludes and dupes have been filtered out by now. */
 	if (isLinkable(flp->fl_mode)) {
 	    if (flp->fl_nlink == 1 || !seenHardLink(fl->files, flp, &fileid)) {
@@ -1204,10 +1216,6 @@ static void genCpioListAndHeader(FileList fl, rpmSpec spec, Package pkg, int isS
 	    headerPutUint32(h, RPMTAG_FILEMTIMES, &rtime, 1);
 	}
 
-	{   rpm_mode_t rmode = (rpm_mode_t) flp->fl_mode;
-	    headerPutUint16(h, RPMTAG_FILEMODES, &rmode, 1);
-	}
-
 	{   rpm_rdev_t rrdev = (rpm_rdev_t) flp->fl_rdev;
 	    headerPutUint16(h, RPMTAG_FILERDEVS, &rrdev, 1);
 	}
@@ -1280,6 +1288,20 @@ static void genCpioListAndHeader(FileList fl, rpmSpec spec, Package pkg, int isS
     }
     pkg->dpaths[npaths] = NULL;
 
+    if (!allModes.empty()) {
+        headerPutUint16(h, RPMTAG_FILEMODES, allModes.data(), allModes.size());
+    }
+    if(!allGenValues.empty()){
+	    headerPutUint32(h, RPMTAG_FILEGENVALUE, allGenValues.data(), allGenValues.size());
+    }
+    if(!allCCSIDs.empty()){
+	    headerPutUint16(h, RPMTAG_FILECCSID, allCCSIDs.data(), allCCSIDs.size());
+    }
+
+    if(!allTextFlags.empty()){
+	    headerPutUint16(h,  RPMTAG_FILETXTFLAG, allTextFlags.data(), allTextFlags.size());
+	}
+
     /* Use 64bit sizes always on v6, on older only if required. */
     if (spec->rpmformat < 6 && totalFileSize < UINT32_MAX) {
 	rpm_off_t totalsize = totalFileSize;
@@ -1404,7 +1426,6 @@ static rpmRC addFile(FileList fl, const char * diskPath,
     const char *fileUname;
     const char *fileGname;
     rpmRC rc = RPMRC_FAIL; /* assume failure */
-
     /* Strip trailing slash. The special case of '/' path is handled below. */
     if (plen > 0 && diskPath[plen - 1] == '/') {
 	diskPath = strcpy(buf, diskPath);
@@ -1488,7 +1509,7 @@ static rpmRC addFile(FileList fl, const char * diskPath,
 	return recurseDir(fl, diskPath);
     }
 
-    fileMode = statp->st_mode;
+    fileMode = statp->st_mode & (S_IFMT | 07777);
 
     /* Explicit %attr() always wins */
     if (fl->cur.ar.ar_fmodestr) {
@@ -1685,7 +1706,9 @@ static rpmRC processMetadataFile(Package pkg, FileList fl,
     rc = RPMRC_OK;
 
     if (absolute)
+    {
 	rc = addFile(fl, fn, NULL);
+    }
 
 exit:
     free(apkt);
@@ -2201,7 +2224,9 @@ static rpmRC processBinaryFile(Package pkg, FileList fl, const char * fileName,
 	goto exit;
 
     for (i = 0; i < argc; i++)
+    {
 	rc = addFile(fl, argv[i], NULL);
+    }
     argvFree(argv);
 
 exit:
@@ -2615,7 +2640,6 @@ static rpmRC processPackageFiles(rpmSpec spec, rpmBuildPkgFlags pkgFlags,
     /* Verify that file attributes scope over hardlinks correctly. */
     if (checkHardLinks(fl.files) && spec->rpmformat < 6)
 	(void) rpmlibNeedsFeature(pkg, "PartialHardlinkSets", "4.0.4-1");
-
     genCpioListAndHeader(&fl, spec, pkg, 0);
 
 exit:
