diff --git a/lib/fprint.cc b/lib/fprint.cc
index 0a7df8d..9041bee 100644
--- a/lib/fprint.cc
+++ b/lib/fprint.cc
@@ -18,6 +18,7 @@
 #include "fprint.hh"
 #include "misc.hh"
 #include "debug.h"
+#include "rpmchroot.hh"
 #include <libgen.h>
 
 /**
@@ -166,14 +167,26 @@ static int doLookupId(fingerPrintCache cache,
 	cacheHit = cacheContainsDirectory(cache, fpId);
 	if (cacheHit != NULL) {
 	    fp->entry = cacheHit;
-	} else if (!stat(rpmstrPoolStr(cache->pool, fpId), &sb)) {
-	    struct fprintCacheEntry_s newEntry = {
-		.dirId = fpId,
-		.dev = sb.st_dev,
-		.ino = sb.st_ino,
-	    };
-	    auto ret = cache->ht.insert({fpId, newEntry});
-	    fp->entry = &ret->second;
+	} else {
+	    char *statPath = NULL;
+	    const char *rootDir = rpmChrootPath();
+	    const char *dirStr = rpmstrPoolStr(cache->pool, fpId);
+	    if (rootDir && getuid() != 0) {
+		statPath = rpmGenPath(rootDir, dirStr, NULL);
+	    } else {
+		statPath = xstrdup(dirStr);
+	    }
+	    int rc = stat(statPath, &sb);
+	    free(statPath);
+	    if (!rc) {
+		struct fprintCacheEntry_s newEntry = {
+		    .dirId = fpId,
+		    .dev = sb.st_dev,
+		    .ino = sb.st_ino,
+		};
+		auto ret = cache->ht.insert({fpId, newEntry});
+		fp->entry = &ret->second;
+	    }
 	}
 
         if (fp->entry) {
