diff --git a/lib/rpmrc.cc b/lib/rpmrc.cc
index 2b3baed..95fbacb 100644
--- a/lib/rpmrc.cc
+++ b/lib/rpmrc.cc
@@ -1364,6 +1364,12 @@ static void defaultMachine(rpmrcCtx ctx, const char ** arch, const char ** os)
 	}
 #	endif
 
+# if defined(__MVS__)
+  {
+        strcpy(un.machine, "s390x");
+        strcpy(un.sysname, "zos");
+  }
+#endif
 	/* the uname() result goes through the arch_canon table */
 	const canonEntry * canon = lookupInCanonTable(un.machine,
 			   ctx->tables[RPM_MACHTABLE_INSTARCH].canons);
