diff --git a/lib/keystore.cc b/lib/keystore.cc
index 57a8c71..70bf7aa 100644
--- a/lib/keystore.cc
+++ b/lib/keystore.cc
@@ -23,6 +23,8 @@
 
 #include "debug.h"
 
+#include <zos-filesystem.h>
+
 using std::string;
 using namespace rpm;
 namespace fs = std::filesystem;
@@ -118,15 +120,21 @@ exit:
 static rpmRC delete_file_store(std::string path)
 {
     try {
+#ifdef __MVS__
+	portable_remove_all(path);
+#else
 	fs::remove_all(path);
-    } catch (std::filesystem::filesystem_error& e) {
+#endif
+    } catch (const std::exception& e) {
+	rpmlog(RPMLOG_ERR, _("Failed to remove keystore at %s: %s\n"),
+	       path.c_str(), e.what());
 	return RPMRC_FAIL;
     }
     return RPMRC_OK;
 }
 
 /* Wrapper for private delete_key() methods that falls back to short keyid */
-static rpmRC rpm::delete_key_compat(auto keystore, rpmtxn txn, rpmPubkey key, auto skip)
+rpmRC rpm::delete_key_compat(auto keystore, rpmtxn txn, rpmPubkey key, auto skip)
 {
     rpmRC rc = RPMRC_NOTFOUND;
     string fp = rpmPubkeyFingerprintAsHex(key);
@@ -223,8 +231,9 @@ static int acquire_write_lock(rpmtxn txn)
 }
 
 static void free_write_lock(int fd)
-{
+{ 
     flock(fd, LOCK_UN);
+    close(fd);
 }
 
 rpmRC keystore_openpgp_cert_d::load_keys(rpmtxn txn, rpmKeyring keyring)
