diff --git a/lib/keystore.cc b/lib/keystore.cc
index 4eaca9c..ee66146 100644
--- a/lib/keystore.cc
+++ b/lib/keystore.cc
@@ -23,6 +23,8 @@
 
 #include "debug.h"
 
+#include <zos-filesystem.h>
+
 using std::string;
 using namespace rpm;
 namespace fs = std::filesystem;
@@ -117,7 +119,8 @@ exit:
 static rpmRC delete_file_store(std::string path)
 {
     try {
-	fs::remove_all(path);
+	// fs::remove_all(path);
+	portable_remove_all(path);
     } catch (std::filesystem::filesystem_error& e) {
 	return RPMRC_FAIL;
     }
@@ -199,7 +202,9 @@ static int acquire_write_lock(rpmtxn txn)
 
     if ((fd = open(lockpath, O_WRONLY|O_CREAT, 644)) == -1) {
 	rpmlog(RPMLOG_ERR, _("Can't create writelock for keyring at %s: %s\n"), keyringpath, strerror(errno));
+    #ifndef __MVS__
     } else if (flock(fd, LOCK_EX|LOCK_NB)) {
+    #endif 
 	rpmlog(RPMLOG_ERR, _("Can't acquire writelock for keyring at %s\n"), keyringpath);
 	close(fd);
 	fd = -1;
@@ -212,8 +217,10 @@ static int acquire_write_lock(rpmtxn txn)
 }
 
 static void free_write_lock(int fd)
-{
+{ 
+#ifndef __MVS__
     flock(fd, LOCK_UN);
+#endif
 }
 
 rpmRC keystore_openpgp_cert_d::load_keys(rpmtxn txn, rpmKeyring keyring)
