diff --git a/misc/system.h b/misc/system.h
index 596bec0..81b286f 100644
--- a/misc/system.h
+++ b/misc/system.h
@@ -82,4 +82,15 @@ extern int fdatasync(int fildes);
 
 #define N_(Text) Text
 
+#if !defined(HAVE_STPCPY) && defined(__MVS__)
+inline char *stpcpy(char *dest, const char *src)
+{
+    while ((*dest = *src) != '\0') {
+        dest++;
+        src++;
+    }
+    return dest;
+}
+#endif
+
 #endif	/* H_SYSTEM */
