diff --git a/misc/system.h b/misc/system.h
index 36ea094..e036964 100644
--- a/misc/system.h
+++ b/misc/system.h
@@ -77,7 +77,11 @@ extern int fdatasync(int fildes);
 
    setprogname(*pn) must be the first call in main() in order to set the name
    as soon as possible. */
-#if defined(HAVE_SETPROGNAME) /* BSD'ish systems */
+#if defined(__MVS__)
+  #define setprogname(pn) ((void)0)
+#endif
+
+#if defined(HAVE_SETPROGNAME) || defined(__MVS__)/* BSD'ish systems */
 # include <stdlib.h> /* Make sure this header is included */
 # define xsetprogname(pn) setprogname(pn)
 # define xgetprogname(pn) getprogname(pn)
@@ -100,4 +104,15 @@ extern int fdatasync(int fildes);
 
 #define N_(Text) Text
 
+#if !defined(HAVE_STPCPY) && defined(__MVS__)
+inline char *stpcpy(char *dest, const char *src)
+{
+    while ((*dest = *src) != '\0') {
+        dest++;
+        src++;
+    }
+    return dest;
+}
+#endif
+
 #endif	/* H_SYSTEM */
