diff --git a/misc/fts.c b/misc/fts.c
index 59e0731..ccaf0bb 100644
--- a/misc/fts.c
+++ b/misc/fts.c
@@ -37,7 +37,7 @@ static char sccsid[] = "@(#)fts.c	8.6 (Berkeley) 8/14/94";
 #endif
 
 /* Conditional to set up proper fstat64 implementation */
-#if defined(hpux) || defined(sun) || (defined(__APPLE__) && defined(_DARWIN_FEATURE_ONLY_64_BIT_INODE))
+#if defined(hpux) || defined(sun) || (defined(__APPLE__) && defined(_DARWIN_FEATURE_ONLY_64_BIT_INODE)) || defined(__MVS__)
 #   define FTS_FSTAT64(_fd, _sbp)   fstat((_fd), (_sbp))
 #else
 #   define FTS_FSTAT64(_fd, _sbp)   fstat64((_fd), (_sbp))
@@ -56,19 +56,21 @@ static char sccsid[] = "@(#)fts.c	8.6 (Berkeley) 8/14/94";
 #else
 
 /* Conditionals for working around non-GNU environments */
-#if defined(hpux)
+#if defined(hpux) || defined(__MVS__)
 #   define        _INCLUDE_POSIX_SOURCE
 #   define __errno_location() 	(&errno)
-#   define dirfd(dirp)		-1
 #   define stat64		stat
 #endif
+#if defined(hpux)
+#   define dirfd(dirp)		-1
+#endif
 #if defined(sun)
 #   define __errno_location()	(&errno)
 #   define dirfd(dirp)		-1
 #endif
 #if defined(__APPLE__)
 #   define __errno_location()	(__error())
-#if defined(_DARWIN_FEATURE_ONLY_64_BIT_INODE)
+#if defined(_DARWIN_FEATURE_ONLY_64_BIT_INODE) || defined(__MVS__)
 #		define stat64	stat
 #endif
 #endif
@@ -83,7 +85,8 @@ static char sccsid[] = "@(#)fts.c	8.6 (Berkeley) 8/14/94";
 #include <dirent.h>
 #include <errno.h>
 #include "rpmfts.h"
-#   define __set_errno(val) (*__errno_location ()) = (val)
+//  #   define __set_errno(val) (*__errno_location ()) = (val)
+#define __set_errno(val) (errno = (val))
 #   define __open	open
 #   define __close	close
 #   define __fchdir	fchdir
@@ -313,7 +316,8 @@ Fts_close(FTS * sp)
 		if (saved_errno != 0) {
 			/* Free up the stream pointer. */
 			free(sp);
-			__set_errno (saved_errno);
+		//	__set_errno (saved_errno);
+			errno=saved_errno;
 			return (-1);
 		}
 	}
@@ -1132,7 +1136,11 @@ static int
 fts_safe_changedir(FTS * sp, FTSENT * p, int fd, const char * path)
 {
 	int ret, oerrno, newfd;
+#if defined(__MVS__)
+	struct stat sb;
+#else
 	struct stat64 sb;
+#endif
 
 	newfd = fd;
 	if (ISSET(FTS_NOCHDIR))
