diff --git a/rpmio/rpmlog.cc b/rpmio/rpmlog.cc
index b3a2806..b676f54 100644
--- a/rpmio/rpmlog.cc
+++ b/rpmio/rpmlog.cc
@@ -259,7 +259,9 @@ static int getColorConfig(void)
 
 static void logerror(void)
 {
-    static __thread int lasterr = 0;
+    // static __thread int lasterr = 0;
+    // assuming thread safety isn't critical
+    static int lasterr = 0;
     if (errno != EPIPE && errno != lasterr) {
 	lasterr = errno;
 	perror(_("Error writing to log"));
@@ -269,7 +271,8 @@ static void logerror(void)
 static int rpmlogDefault(FILE *stdlog, rpmlogRec rec)
 {
     FILE *msgout = (stdlog ? stdlog : stderr);
-    static __thread int color = -1;
+    //static __thread int color = -1;
+    static int color = -1;
     const char * colorOn = NULL;
 
     if (color < 0)
