diff --git a/rpmio/macro.cc b/rpmio/macro.cc
index 21d1757..07d2530 100644
--- a/rpmio/macro.cc
+++ b/rpmio/macro.cc
@@ -1131,19 +1131,24 @@ static void doShescape(rpmMacroBuf mb, rpmMacroEntry me, ARGV_t argv, size_t *pa
 	rpmMacroBufAppend(mb, '\'');
     }
 }
-
 static uint64_t getmem_total(void)
 {
     uint64_t mem = 0;
+
+#ifdef __MVS__
+    /* z/OS: Use CVTRLSTG to get actual real storage online at IPL in K */
+    mem = (uint64_t)((int)((char *__ptr32 *__ptr32 *)0)[4][214]) * 1024;
+#else
     long int pagesize = sysconf(_SC_PAGESIZE);
     long int pages = sysconf(_SC_PHYS_PAGES);
 
     if (pagesize <= 0)
-	pagesize = 4096;
+       pagesize = 4096;
     if (pages > 0) {
-	/* Cast needed to force 64bit calculation on 32bit systems */
-	mem = (uint64_t)pages * pagesize;
+       /* Cast needed to force 64bit calculation on 32bit systems */
+       mem = (uint64_t)pages * pagesize;
     }
+#endif
 
     return mem;
 }
