diff --git a/rpmio/rpmpgp_dummy.cc b/rpmio/rpmpgp_dummy.cc
index 97babb6..280d8b5 100644
--- a/rpmio/rpmpgp_dummy.cc
+++ b/rpmio/rpmpgp_dummy.cc
@@ -121,3 +121,11 @@ rpmRC pgpPubkeyMerge(const uint8_t *pkts1, size_t pkts1len, const uint8_t *pkts2
 {
     return RPMRC_FAIL;
 }
+
+int pgpDigParamsSalt(pgpDigParams digp, const uint8_t **datap, size_t *lenp)
+{
+    (void)digp;
+    if (datap) *datap = NULL;
+    if (lenp)  *lenp = 0;
+    return -1;
+}
