diff --git a/rpmio/rpmlua.cc b/rpmio/rpmlua.cc
index 9dde5f0..0847a03 100644
--- a/rpmio/rpmlua.cc
+++ b/rpmio/rpmlua.cc
@@ -1204,7 +1204,7 @@ static int rpm_open(lua_State *L)
     FD_t fd = Fopen(path, mode);
 
     if (fd == NULL) {
-	return luaL_error(L, "%s open failed: %s", path, strerror(errno));
+	return luaL_error(L, "%s open failed rpmio_rpmlua.cc: %s", path, strerror(errno));
     }
 
     return newinstance(L, "rpm.fd", fd);
