diff --git a/scripts/brp-remove-la-files b/scripts/brp-remove-la-files
index 0e6e58f..a45f292 100755
--- a/scripts/brp-remove-la-files
+++ b/scripts/brp-remove-la-files
@@ -11,6 +11,6 @@ do
 
   find "$RPM_BUILD_ROOT" -type $FILETYPE -name '*.la' 2>/dev/null -print0 |
     xargs -0 grep --fixed-strings '.la - a libtool library file' --files-with-matches --null |
-    xargs -0 rm --force
+    xargs -0 rm -f
 
 done
diff --git a/scripts/brp-strip b/scripts/brp-strip
index 76b360b..bc4fd7e 100755
--- a/scripts/brp-strip
+++ b/scripts/brp-strip
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/env bash
 
 # If using normal root, avoid changing anything.
 if [ -z "$RPM_BUILD_ROOT" ] || [ "$RPM_BUILD_ROOT" = "/" ]; then
diff --git a/scripts/check-files b/scripts/check-files
index 12bacfb..f0cc8c9 100755
--- a/scripts/check-files
+++ b/scripts/check-files
@@ -23,7 +23,7 @@ fi
 FILES_DISK=`mktemp "${TMPDIR}/rpmXXXXXX"`
 
 # Ensure temporary file is cleaned up when we exit
-trap "rm -f \"${FILES_DISK}\"" 0 2 3 5 10 13 15
+trap "rm -f \"${FILES_DISK}\"" 0 INT QUIT TRAP USR1 PIPE TERM
 
 # Find non-directory files in the build root and compare to the manifest.
 # TODO: regex chars in last sed(1) expression should be escaped
