diff --git a/scripts/brp-remove-la-files b/scripts/brp-remove-la-files
index 0e6e58f..a45f292 100755
--- a/scripts/brp-remove-la-files
+++ b/scripts/brp-remove-la-files
@@ -11,6 +11,6 @@ do
 
   find "$RPM_BUILD_ROOT" -type $FILETYPE -name '*.la' 2>/dev/null -print0 |
     xargs -0 grep --fixed-strings '.la - a libtool library file' --files-with-matches --null |
-    xargs -0 rm --force
+    xargs -0 rm -f
 
 done
diff --git a/scripts/brp-strip b/scripts/brp-strip
index 76b360b..47840b6 100755
--- a/scripts/brp-strip
+++ b/scripts/brp-strip
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/env bash
 
 # If using normal root, avoid changing anything.
 if [ -z "$RPM_BUILD_ROOT" ] || [ "$RPM_BUILD_ROOT" = "/" ]; then
@@ -14,7 +14,7 @@ NCPUS=${RPM_BUILD_NCPUS:-1}
 MAX_ARGS=32
 
 case `uname -a` in
-Darwin*) exit 0 ;;
+Darwin|OS/390*) exit 0 ;;
 *) ;;
 esac
 
diff --git a/scripts/brp-strip-comment-note b/scripts/brp-strip-comment-note
index 982b6c0..c40d8e3 100755
--- a/scripts/brp-strip-comment-note
+++ b/scripts/brp-strip-comment-note
@@ -9,7 +9,7 @@ OBJDUMP=${2:-objdump}
 NCPUS=${RPM_BUILD_NCPUS:-1}
 
 case `uname -a` in
-Darwin*) exit 0 ;;
+Darwin*|OS/390*) exit 0 ;;
 *) ;;
 esac
 
diff --git a/scripts/brp-strip-static-archive b/scripts/brp-strip-static-archive
index be869fa..5386793 100755
--- a/scripts/brp-strip-static-archive
+++ b/scripts/brp-strip-static-archive
@@ -8,7 +8,7 @@ STRIP=${1:-strip}
 NCPUS=${RPM_BUILD_NCPUS:-1}
 
 case `uname -a` in
-Darwin*) exit 0 ;;
+Darwin|OS/390*) exit 0 ;;
 *) ;;
 esac
 
diff --git a/scripts/check-files b/scripts/check-files
index 12bacfb..f0cc8c9 100755
--- a/scripts/check-files
+++ b/scripts/check-files
@@ -23,7 +23,7 @@ fi
 FILES_DISK=`mktemp "${TMPDIR}/rpmXXXXXX"`
 
 # Ensure temporary file is cleaned up when we exit
-trap "rm -f \"${FILES_DISK}\"" 0 2 3 5 10 13 15
+trap "rm -f \"${FILES_DISK}\"" 0 INT QUIT TRAP USR1 PIPE TERM
 
 # Find non-directory files in the build root and compare to the manifest.
 # TODO: regex chars in last sed(1) expression should be escaped
