diff --git a/sign/rpmgensig.cc b/sign/rpmgensig.cc
index 4f3618e..e029200 100644
--- a/sign/rpmgensig.cc
+++ b/sign/rpmgensig.cc
@@ -78,7 +78,7 @@ static int manageFile(FD_t *fdp, const char *fn, int flags)
 	}
 	fd = Fopen(fn, fmode);
 	if (fd == NULL || Ferror(fd)) {
-	    rpmlog(RPMLOG_ERR, _("%s: open failed: %s\n"), fn,
+	    rpmlog(RPMLOG_ERR, _("%s: sign/rpmgensig.cc open failed: %s\n"), fn,
 		Fstrerror(fd));
 	    return 1;
 	}
